File: solutions/typescript/anagram/1/anagram.ts
export class Anagram {
  private input: string;
  private sorted: string;

  constructor(input: string) {
    this.input = input.toLowerCase();
    this.sorted = [...this.input].sort().join("");
  }

  public matches(...potentials: string[]): string[] {
    return potentials.filter((word) => {
      const w = word.toLowerCase();

      if (w == this.input) return false;
      if (w.length != this.input.length) return false;

      return [...w].sort().join("") == this.sorted;
    });
  }
}

File: solutions/elixir/anagram/1/lib/anagram.ex
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t(), [String.t()]) :: [String.t()]
  def match(base, candidates) do
    base = to_downcase_list(base)
    Enum.filter(candidates, &anagram?(to_downcase_list(&1), base))
  end

  defp anagram?(word, base) when word == base, do: false
  defp anagram?(word, base), do: Enum.sort(word) == Enum.sort(base)

  defp to_downcase_list(word) do
    word |> String.downcase() |> String.graphemes()
  end
end
